File: leetcode/01-matrix.py
from collections import deque
from typing import List


class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        n, m = len(mat), len(mat[0])
        dirs = [(-1, 0), (0, -1), (1, 0), (0, 1)]
        res = [[0] * m for _ in range(n)]

        queue = deque()
        prev = {}

        for i in range(0, n):
            for j in range(0, m):
                if mat[i][j] == 0:
                    queue.append((i, j))
                    prev[(i, j)] = 0

        while queue:
            row, col = queue.popleft()

            for d in dirs:
                r = row + d[0]
                c = col + d[1]
                if r >= 0 and r < n and c >= 0 and c < m and (r, c) not in prev:
                    prev[(r, c)] = prev[(row, col)] + 1
                    queue.append((r, c))
                    res[r][c] = prev[(r, c)]

        return res

File: leetcode/01-matrix.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat)
    {
        int row = mat.size(), col = mat[0].size();

        function<int(int, int)> get = [&](int r, int c) {
            if (r >= 0 && r < row && c >= 0 && c < col) {
                return mat[r][c];
            }
            return 1000000000;  // see constraints
        };

        // top -> down
        // left -> right
        for (int i = 0; i < row; ++i) {
            for (int j = 0; j < col; ++j) {
                if (mat[i][j]) {
                    mat[i][j] = 1 + min(get(i - 1, j), get(i, j - 1));
                }
            }
        }

        // down -> top
        // right -> left
        for (int i = row - 1; i >= 0; --i) {
            for (int j = col - 1; j >= 0; --j) {
                if (mat[i][j]) {
                    mat[i][j] = min(mat[i][j], 1 + min(get(i + 1, j), get(i, j + 1)));
                }
            }
        }

        return mat;
    }
};
